lib/vm.js

```diff
@@ -390,17 +391,26 @@ function compileFunction(code, params, options = {}) {
     }
   });
 
-  return _compileFunction(
+  const fn = _compileFunction(
     code,
     filename,
     lineOffset,
     columnOffset,
     cachedData,
-    produceCachedData,
     parsingContext,
     contextExtensions,
     params
   );
+
+  if (produceCachedData) {
```

Comment:
i would rather the could just call this manually. we've learned from vm.Script that `produceCachedData` isn't a very good model

Reply:
@devsnek that could be done. A couple of things:

1. That'll force us to make this an API method. `vm.compileFunction` used to do that, and people should be allowed to fallback to the original without breakage.

2. That'll be a breaking change for `vm.compileFunction`.

Perhaps I could do that in a separate `semver-major` PR? That said, I'm still not 100% convinced that this would be a good idea in the first place.

Reply:
The returned values of vm.compileFunction are neither documented or tested, and it’s only been released since 10.11 (10.10?) I would say it should probably be safe to deprecate the option in 11

Reply:
Also given how the returned values are used, does it actually break anyone if the cache is not generated? If we only return the function as-is, then cachedDataProduced and cachedData will always be undefined which is not that different from “the cache generation always fail”, but the users (if they exist) should always prepare to handle that case in the first place (provided they actually guess what those values are even when there are no documentations)

Reply:
(The comments above are orthogonal to this PR, BTW, but if vm.compileFunction does not use it, and it’s not exposed as an API either, I couldn’t see why we are adding a binding that is not used...)

Reply:
The key question is how we should handle the absence of cached data. Should it be seen as an error or a normal state? My inclination is to treat it as a normal state for non-tooling use cases, as it's unclear how extensively these values have been used in production code.

I support the idea of a separate semver-major PR to address this and consider making produceCachedData an API method. This approach provides clarity and a smoother transition for users, especially in scenarios where the use of these values is minimal.